### PR TITLE
Temporary fix of different memory formats between Pytorch and TOSA

### PR DESCRIPTION
**Pull request by Jerry-Ge**
- This patch will be reverted once Github issue #1110 is resolved
- Torch graph operators like Conv2d are using NCHW formats while the TOSA equivalents are using NHWC formats.
- In the previous implementation, a pair of transpose operators are added around the TOSA Conv2d operator to cope with the memory format differences.
- In this patch, it removes those intermediate transposes around operators like Conv2d.
- Instead, we're transposing the inputs and outputs around the whole lowered graph and the Tosa compiler assumes the torch graph is in NHWC format so all the input/output shapes can match.

### CODE REVIEW

File: backends/arm/test/test_tosa.py [hunk 1]
```diff
@@ -82,19 +82,28 @@ def prepare_model_and_ref(test_model, profile=TosaProfile.MI):
...
         captured_model_graph_module = capture_pre_autograd_graph(
             model, copy.deepcopy(model.inputs[profile])
         )
         # Setup the quantizer
         quantizer = XNNPACKQuantizer()
-        operator_config = get_symmetric_quantization_config(is_per_channel=False)
+
+        # TODO: remove this hack, haven't figure out entirely but the int8 mnv2
+        # can only pass if we set the weight_qmin=0 here
+        weight_qmin = 0 if test_model == "mobilenetv2" else -127
```
Comment (Jerry-Ge) [inline backends/arm/test/test_tosa.py:95]: we can remove this line if we set `quantize(Quantize(calibrate=False))` like [here](https://github.com/pytorch/executorch/blob/42eeebc7899295d76a235eab4563efd38c1feb98/backends/xnnpack/test/models/mobilenet_v2.py#L55)

File: backends/arm/test/models/test_mobilenet_v2_arm.py [hunk 2]
```diff
@@ -30,37 +30,36 @@ class TestMobileNetV2(unittest.TestCase):
...
                 inputs=self.model_inputs,
                 profile=TosaProfile.BI,
                 backend=ArmBackendSelector.TOSA,
+                permute_memory_to_nhwc=True,
             )
             .quantize()
             .export()
-            .check(list(self.all_operators))
             .to_edge()
+            .check(list(self.all_operators))
```
Comment (Jerry-Ge) [inline backends/arm/test/models/test_mobilenet_v2_arm.py:62]: this needs updated to remove the check of `executorch_exir_dialects_edge__ops_aten__native_batch_norm_legit_no_training_default`

File: backends/arm/test/tosautil/tosa_test_utils.py [hunk 3]
```diff
@@ -164,6 +165,11 @@ def run_tosa_ref_model(
             params_input[0], params_input[1], inputs
         ):
             data_np = data.detach().numpy()
+
+            if permute_memory_to_nhwc:
+                NHWC_Order = (0, 2, 3, 1)
+                data_np = np.transpose(data_np, NHWC_Order)
```
Comment (digantdesai) [inline backends/arm/test/tosautil/tosa_test_utils.py:171]: Shouldn't this transpose be done outside this method? I.e. this method takes in whatever tensor and runs the model. If it happens to be NHWC that is ok.

File: backends/arm/tosa_quant_utils.py [hunk 4]
```diff
@@ -196,7 +196,15 @@ def build_rescale_conv_output(
     post_conv2d_scale = (input_scale.number * weight_scale.number) / output_scale.number
 
     # Since we assume the input tensor that is being rescaled is int32 date type, zero point must be 0.
-    rescale_op = build_rescale(
-        tosa_fb, post_conv2d_scale, op, output_type, op.shape, 0, 0, double_round
+    build_rescale(
```
Comment (digantdesai) [inline backends/arm/tosa_quant_utils.py:199]: Unrelated to NHWC changes?

File: backends/arm/test/tester/arm_tester.py [hunk 5]
```diff
@@ -61,20 +62,26 @@ def __init__(
...
+
         if backend == ArmBackendSelector.TOSA:
             self.tosa_test_util = TosaTestUtils(profile=profile)
             # The spec below tiggers arm_backend.py to output two files:
             #   1) output.tosa
             #   2) desc.json
             # Saved on disk in self.tosa_test_util.intermediate_path
             self.compile_spec = generate_tosa_compile_spec(
-                self.tosa_test_util.intermediate_path
+                self.permute_memory_to_nhwc, self.tosa_test_util.intermediate_path
```
Comment (digantdesai) [inline backends/arm/test/tester/arm_tester.py:79]: In the long run imagine this to be transparent to the delegate logic. I.e all the delegate does is add `converts` if required. But if you have a graph like,

G1 `input --> nhwc_only_op --> output`
G2 `input --> convert_to_nhwc --> nhwc_only_op --> convert_to_nhwc --> output` added by a pass before `to_backed` or first thing inside `arm_delegate::preprocess()`.
G3 `input --> arm_delegate_op --> output`

File: backends/arm/test/models/test_mobilenet_v2_arm.py [hunk 6]
```diff
@@ -30,37 +31,39 @@ class TestMobileNetV2(unittest.TestCase):
         "executorch_exir_dialects_edge__ops_aten_convolution_default",
     }
 
-    @unittest.skip("This test is not supported yet")
+    operators_after_quantization = all_operators - {
```
Comment (digantdesai) [inline backends/arm/test/models/test_mobilenet_v2_arm.py:34]: Unrelated changes? Please move them to a different PR?

File: backends/arm/operators/op_conv2d.py [hunk 7]
```diff
@@ -59,12 +54,6 @@ def define_node(
         # Currently only int8 is supported in quantized types.
         actual_out_type = ts.DType.INT8 if is_quant_node else output.dtype
 
-        ## Transpose input tensor to NHWC_Order for TOSA
```
Comment (digantdesai) [inline backends/arm/operators/op_conv2d.py:62]: Do we want to assert the dim order here?

File: backends/arm/operators/op_common.py [hunk 8]
```diff
@@ -22,22 +21,10 @@ def build_avg_pool_2d_common(
...
 ):
     accumulator_type = input_tensor.dtype
-    input_output_type = input_tensor.dtype
 
     if is_quant_node:
         # Accumulator type always is int32 when input tensor is an integer type.
         accumulator_type = ts.DType.INT32
-        input_output_type = ts.DType.INT8
-
-    # Torch's input is [N,C,H,W], TOSA is [N, H, W, C],
```
Comment (digantdesai) [inline backends/arm/operators/op_common.py:32]: same here, check its really NHWC

File: backends/arm/operators/op_add.py [hunk 9]
```diff
@@ -51,6 +52,13 @@ def define_node(
             inputA_rescale_scale = input_A_scale.number / max_scale_2x
             inputB_rescale_scale = input_B_scale.number / max_scale_2x
 
+            broadcasted_shape = broadcast_shapes(input_A.shape, input_B.shape)
+            if permute_memory_to_nhwc:
```
Comment (digantdesai) [inline backends/arm/operators/op_add.py:56]: if you insert a convert node in the Graph and retrace it, it should convert all the intermediate tensors to NHWC format, can you check that so you just have to assert and not convert foreach op.

File: backends/arm/arm_backend.py [hunk 10]
```diff
@@ -63,14 +66,17 @@ def generate_ethosu_compile_spec(
 
 
 def generate_tosa_compile_spec(
+    permute_memory_to_nhwc: Optional[bool] = None,
```
Comment (digantdesai) [inline backends/arm/arm_backend.py:69]: why is this a choice? i.e. don't we always want this to be true?

File: backends/arm/arm_backend.py [hunk 11]
```diff
@@ -138,13 +156,28 @@ def preprocess(  # noqa: C901
 
                 # Visiting each Node
                 if node.target.__name__ in node_visitors:
-                    node_visitors[node.target.__name__].define_node(
-                        node, tosa_graph, inputs, output, is_quant_node(node)
-                    )
+                    if node.target.__name__ in [
+                        "aten.add.Tensor",
+                        "aten._native_batch_norm_legit_no_training.default",
```
Comment (digantdesai) [inline backends/arm/arm_backend.py:161]: Other ops can handle !nhwc?

File: backends/arm/arm_backend.py (hunk 10 above)
Comment (Jerry-Ge) [reply backends/arm/arm_backend.py:69]: i agree. my original thought was we could also permute to other memory_formats but this seems less useful.

File: backends/arm/arm_backend.py (hunk 11 above)
Comment (Jerry-Ge) [reply backends/arm/arm_backend.py:161]: yes, for each test case right now, we're annotating them either to do `permute to nhwc` or not by default

File: backends/arm/test/models/test_mobilenet_v2_arm.py (hunk 6 above)
Comment (Jerry-Ge) [reply backends/arm/test/models/test_mobilenet_v2_arm.py:34]: i would suggest put it here since for the following reasons: 
- to avoid unnecessary logistical overheads
- this patch's main goal is to support MNV2, we need this change for MNV2 unittest case

File: backends/arm/test/tosautil/tosa_test_utils.py (hunk 3 above)
Comment (Jerry-Ge) [reply backends/arm/test/tosautil/tosa_test_utils.py:171]: agree but i suggest left it here for now due to: 
- we have the `permute_memory_to_nhwc` flag to annotate should it happen or not here as a temporary fix. 
-  this patch will likely be reverted in the long term once Songhao's fix is done. Add a simple check here instead of a new function is a little bit easier to maintain.

File: backends/arm/tosa_quant_utils.py (hunk 4 above)
Comment (Jerry-Ge) [reply backends/arm/tosa_quant_utils.py:199]: yes, it's not very related but as a code improvement for conv2d which is used in MNV2. I would suggest to leave it here (same reasons above)

File: backends/arm/test/tester/arm_tester.py (hunk 5 above)
Comment (Jerry-Ge) [reply backends/arm/test/tester/arm_tester.py:79]: ack.

File: backends/arm/operators/op_add.py (hunk 9 above)
Comment (Jerry-Ge) [reply backends/arm/operators/op_add.py:56]: I'm not too sure how to do this `insert a convert node in the Graph and retrace it` do you have any examples?

File: backends/arm/operators/op_conv2d.py (hunk 7 above)
Comment (Jerry-Ge) [reply backends/arm/operators/op_conv2d.py:62]: do you mean the `dim_order` of input and weight? 
but those orders are still in `NCHW` since they comes from the edge dialect. this fix patch is only assuming they're `NHWC` and is fixing those in the `arm_backend.py` for the output shape and op_placeholders for the input data and shapes.

File: backends/arm/operators/op_add.py (hunk 9 above)
Comment (digantdesai) [reply backends/arm/operators/op_add.py:56]: Just write an `ExportPass`? There are several in XNNPACK preprocess. It should work for memory format, though I haven't tried it. Also it might break with changes Songhao is doing so that is that.

File: backends/arm/arm_backend.py (hunk 11 above)
Comment (digantdesai) [reply backends/arm/arm_backend.py:161]: But for a large network how is this handled when you have let's say
`input -> add* -> sub -> output`? Also why conv is not here?

File: backends/arm/test/tosautil/tosa_test_utils.py (hunk 3 above)
Comment (Jerry-Ge) [reply backends/arm/test/tosautil/tosa_test_utils.py:171]: Fred has a followup patch to improve this as well.

File: backends/arm/test/test_models.py [hunk 12]
```diff
@@ -52,6 +53,8 @@ class simple_clone(torch.nn.Module):
             TosaProfile.MI: (torch.ones(10),),
         }
 
+        permute_memory_to_nhwc = False
```
[the remark this answers is not in the document]
Comment (Jerry-Ge) [reply backends/arm/test/test_models.py:56]: ack.

File: backends/arm/arm_backend.py (hunk 11 above)
Comment (Jerry-Ge) [reply backends/arm/arm_backend.py:161]: > But for a large network how is this handled when you have let's say
using the annotations below. 

> Also for conv is it always set?
yes 

I think this one can also be resolved given the annotation stuff is shown.

File: backends/arm/operators/op_add.py (hunk 9 above)
Comment (Jerry-Ge) [reply backends/arm/operators/op_add.py:56]: yea, looks like a proper fix. how about let's keep my temporary solution now given the deadline and Songhao's on-going fix then we can improve given the suggestion here?

**Comment by Jerry-Ge** [comment]
@guangy10 seems we're missing the tosa_reference_model in the unittest job? it works locally for me and works in ci/trunk. 
@digantdesai could you help do another round of review here or approval? 

thanks guys!

**Review by digantdesai — APPROVED** [review]

**Review by digantdesai — APPROVED** [review]

**Comment by Jerry-Ge** [comment]
thanks @digantdesai for the approval. could we merge this patch now? not sure if everything is all ok from Meta internal checks?

**Comment by digantdesai** [comment]
Rebase please?

**Comment by Jerry-Ge** [comment]
> Rebase please?

thanks @digantdesai , rebase done : )
